File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    private int diameter = 0; 
    public int diameterOfBinaryTree(TreeNode root) {
        helper(root);
        return diameter; }

    private int helper(TreeNode node) {
        if (node == null) return 0; 
        int leftDepth = helper(node.left);  
        int rightDepth = helper(node.right); 
        int currentDiameter = leftDepth + rightDepth;
        diameter = Math.max(diameter, currentDiameter);
        return Math.max(leftDepth, rightDepth) + 1;
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.ans = 0
        def depth(root):
            if not root:
                return 0
            L = depth(root.left)
            R = depth(root.right)
            self.ans = max(self.ans, L + R )
            return max(L, R) + 1
        depth(root)
        return self.ans
        